doc/design/effects_mutability.md

```diff
@@ -32,59 +32,54 @@ made between operating system load versus store effects.
-    `>{}\{}-{}->`).
-- Mutability
-  - `^m t`, `^m t ^t`, `^m t^`: Parametric mutability. `^m` is an subsidiary mutability parameter,
-    whereas `^t` comes after the type name and is the primary mutability parameter. When otherwise
-    unambiguous in a type expression, `t ^t` can be shortened to `t^` (or `t ^`) and the parameter
-    takes the same name as the type.
-  - `^_ t`, `^_ t^_`: Mutability is contextually irrelevant, and therefore both immutable and
-    mutable values are supported.
-  - `t&`: Non-parametric primary mutability. `t&!` indicates a mutation effect on the value, which
-    must therefore be mutable, as in `... -> 'a t&! -> ...`.
-  - `t\&`, `^_ t^_\&`: Transitive prohibited mutability constraint.
-- Parametric types:
-  - `'a`: Parametric type, whether simple as in `uns` or with its own parameters as in
-    `('aa, ^m, >e) 'a^`.
-  - `'a t`, `('a, ^m, >e) t`, ...: Primarily immutable parametric type.
-  - `'a t&`, `('a, ^m, >e) t&`, ...: Primarily mutable parametric type.
-  - `'a& t`: Subsidiarily mutable parametric type.
-  - `'a t ^t`, `'a t^`, `('a, ^m) t^`: Parametric type with parametric primary mutability. `^m` is a
-    subsidiary mutability parameter, whereas the primary mutability parameter `^t` is distinct from
-    the comma-separated `(...)` parameter list, such that its absence and the absence of `&`
-    indicates that the type is primarily immutable.
-  - `val f: 'a -> 'a t^`: The function produces an isolated value with parametric mutability. The
-    value's mutability is either immutable or mutable as determined by type unification in the
+    + `os`: Operating system load-store effect
+    + `ld`: Runtime system load effect
+    + `mut`: Runtime system load-store effect
+    + `hlt`: Runtime system may-halt effect
+  - `val f >e: ... >e-> ...`: Parametric effect.
+  - `val f >e\{os|mut}: ... >e-> ...`: Parametric effects constraint; transitively prohibited effects
+    are enumerated as `\...` following parametric effect.
+  - `(a: type) -> !&t a -> unit`, `(a: type) -> &t !&a >hlt-> unit`: Mutation effect on mutable
+    function parameter type prefixed by `!`. The affected type need not be the outermost type, but
+    the affected type must be mutable.
+  - `(a: type) -> &t a -> &t a`: Load effect on mutable function parameter. A mutable function
+    parameter always implies a `ld` effect on the parameter, without which the parameter would be
+    useless.
+  - `uns >{&t|(!&u)}-> unit`, `!&t >!&t-> unit`: Type-constrained environment `ld`/`mut` effect(s)
+    during function application. Parentheses around type-constrained effects are optional, though
+    they often improve readability. Note that a `!&t` parameter mutation is distinct from
+    `>(!&t)->` environment mutation.
+  - `uns >ld-> unit`, `uns >mut-> unit`: Environment `ld`/`mut` effect during function application.
+    Type-agnostic environment effects impact mutable values of all types.
```

Comment:
As discussed offline: These examples show how a function can mutate values that were captured in a lexical closure. Often the compiler can be more accurate than `ld` or `mut`, but if the mutation is of a module field binding and the lexical closure was from within the module, the type cannot be more specific because that would require recursive module types.